### ISSUE TITLE
[WIP] Decouple the internal `Config` Go struct from the user config file in `.ipfs/conf`

### ISSUE THREAD

**Issue by schomatis** [ipfs/go-ipfs#8925]
### Checklist

- [X] This is a bug report, not a question. Ask questions on [discuss.ipfs.io](https://discuss.ipfs.io).
- [X] I have searched on the [issue tracker](https://github.com/ipfs/go-ipfs/issues?q=is%3Aissue) for my bug.
- [X] I am running the latest [go-ipfs version](https://dist.ipfs.io/#go-ipfs) or have an issue updating.

### Installation method

ipfs-desktop

### Version

```Text
a
```

### Config

```json
a
```

### Description

_This issue is about fixing our internal technical debt without changing external UX. Some internal APIs will need to be broken but that can be minimized in a progressive deployment._

## Problem

We currently map the JSON data of the user configuration file (UCF) in `.ipfs/conf` to our internal [`Config` Go struct](https://github.com/ipfs/go-ipfs/blob/25cc85fa9359f907f348e0c2139f2b535313c56c/config/config.go#L16). This brings many problems:
* System defaults that might change over releases are hard-coded and locked in the UCF without the user realizing it, or even caring about that particular configuration option.
* The UCF is bloated with configuration options not set by the user making it harder to realize which options have been actually changed or added from the system, developer-defined defaults.
* There is no Go native way to mark a variable as unset. Variables always default to a Go value which might be different from the configuration default we may want that value to encode (the typical example is misinterpreting an unset limit with its default integer value of 0, instead of the expected "no limit" option). To avoid exposing the entire `Config` struct we end up relying in patches like the [`json:",omitempty"` option](https://pkg.go.dev/encoding/json#Marshal), but that ends up creating even more confusion as it only just hides the underlying coupling that remains between these two sources of configuration.

The last point in turn has given rise to "optional" variables, [`OptionalInteger` and related](https://github.com/ipfs/go-ipfs/blob/25cc85fa9359f907f348e0c2139f2b535313c56c/config/types.go#L268), to handle the JSON-GO mismatch of unspecified configuration options with the following drawbacks:
* Increased code bloat and associated maintaining costs, especially because we need one of this Go structures for each variable type the configuration encodes: `Flag`, `Strings`, `Priority`, `OptionalDuration`, `OptionalInteger`, `OptionalString`, and the list just keeps growing.
* Its `WithDefault()` API scatters system default values in the places that _access_ these values, instead of centralizing them in the `config` package where they are _defined_. These system defaults end up hard-coded deep in the code base, or we flat-out copy an external default structure field by field like: 

https://github.com/ipfs/go-ipfs/blob/7162a63e9629258bac0b9d135acae4c3c1b9fb33/core/node/libp2p/relay.go#L28-L40

## Proposed solution

We should clearly differentiate between the internal system `Config` Go struct and the UCF in `.ipfs/conf`. **The second should _not_ the JSON-serialized version of the former.** It is a JSON map with a _subset_ of the options that the user overrides in the internal system `Config` Go struct. We should not leak Go implementation details; for the user the only thing that exists is the JSON nested attribute-value maps (only internally do we use the `Config` struct to validate if those attributes are actually system-defined options).

When the node starts it should load the internal `Config` with its default values and only change the options specified in the UCF, nothing else. The UCF should be a bare-bones file after `ipfs init` which in turn makes user environments easier to parse (both for the user and the developer), that contain only the options deliberately changed by the user (instead of the giant blobs of config posted in each submitted issue left for the reader to decipher if anything relevant to the issue has been actually changed).

In that way all that is currently hard-coded in the UCF through the direct serialization of the `Config` struct (which is slowly and painfully being hidden through the cumbersome "optional" variables) will remain only in the `Config` struct as **system-defined defaults** (subject to change with the release version) that are _never_ exposed in the UCF.

Implementation-wise (and this is the WIP part), to avoid major refactorings of the [`Repo` interface](https://github.com/ipfs/go-ipfs/blob/25cc85fa9359f907f348e0c2139f2b535313c56c/repo/repo.go#L21) (which ideally would need to happen because the repo would no longer contain the main config options, just specific overrides) we can progressively deploy changes through a new structure that differentiates the UCF from the `Config`, with the [defaults](https://github.com/ipfs/go-ipfs/blob/83b3ec4a61f086f81e0f75c5ec9d508bf5385b16/config/init.go#L24-L102) left untouched, just avoiding the serialization to the UCF and instead loading the UCF only to change user-specified values in the internal `Config`.

**Comment by lidel** [comment]
Quick thoughts:
- tldr: my initial  reaction is :+1:  your approach [in this branch](https://github.com/ipfs/go-ipfs/compare/schomatis/fix/config/decouple-user-vs-system-config) feels backward-compatible, so low risk, high return
  - In my mind, the main value added here is **decreased maintenance cost** for Stewards and  contributors
    - We no longer need to do additional QA ensuring config changes are backward-compatible and we won't pollute JSON with empty structs just to  avoid panics.
     - **UX is also improved** (right now we are in-between where some things are in JSON produced by `ipfs init`, some are not, and there is no clear convention why)
- Minimal `$IPFS_PATH/config` being only the diff sgtm
  - There would still be things like `Addresses` and `Identity` which we want to solidify during `ipfs init`, but the rest is just noise at this point 
- Keeping implicit defaults in one place also sgtm
  - I can confirm that grepping `Foo.WithDefault(` every time I want to find where the implicit default is defined is tedious. 
    - It is also error-prone: found myself in situation where I defined different default than the one that was already in other place, and had to refactor/bloat code to make sure only one default is defined at higher layers

**Comment by schomatis** [comment]
Thanks for the feedback @lidel. Will move ahead then with this approach prioritizing backward-compatibility.
